test/xds/xds_client_custom_dialopts_test.go

```diff
@@ -0,0 +1,212 @@
+ * limitations under the License.
+ *
+ */
+
+package xds_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"testing"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
+	internalbootstrap "google.golang.org/grpc/internal/xds/bootstrap"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/xds/bootstrap"
+
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+)
+
+const testCredsBuilderName = "test_dialer_creds"
+
+// testCredsBuilder implements the `Credentials` interface defined in
+// package `xds/bootstrap` and encapsulates an insecure credential with a
+// custom Dialer that specifies how to dial the xDS server.
+type testCredsBuilder struct {
+	dialerCalled     atomic.Bool
+	tagRPCCalled     atomic.Bool
+	handleRPCCalled  atomic.Bool
+	tagConnCalled    atomic.Bool
+	handleConnCalled atomic.Bool
+}
+
+func (t *testCredsBuilder) Build(config json.RawMessage) (credentials.Bundle, func(), error) {
+	cfg := &struct {
+		MgmtServerAddress string `json:"mgmt_server_address"`
+	}{}
+	if err := json.Unmarshal(config, &cfg); err != nil {
+		return nil, func() {}, fmt.Errorf("failed to unmarshal config: %v", err)
+	}
+	return &testCredsBundle{insecure.NewBundle(), &t.dialerCalled, cfg.MgmtServerAddress, &noopStatsHandler{
+		tagRPCCalled:     &t.tagRPCCalled,
+		handleRPCCalled:  &t.handleRPCCalled,
+		tagConnCalled:    &t.tagConnCalled,
+		handleConnCalled: &t.handleConnCalled,
+	}}, func() {}, nil
+}
+
+func (t *testCredsBuilder) Name() string {
+	return testCredsBuilderName
+}
+
+// noopStatsHandler implements `stats.Handler`. It's a no-op mock handler.
+type noopStatsHandler struct {
+	tagRPCCalled     *atomic.Bool
+	handleRPCCalled  *atomic.Bool
+	tagConnCalled    *atomic.Bool
+	handleConnCalled *atomic.Bool
+}
+
+func (h *noopStatsHandler) TagRPC(ctx context.Context, i *stats.RPCTagInfo) context.Context {
+	h.tagRPCCalled.Store(true)
+	return ctx
+}
+
+func (h *noopStatsHandler) HandleRPC(ctx context.Context, s stats.RPCStats) {
+	h.handleRPCCalled.Store(true)
+}
+
+func (h *noopStatsHandler) TagConn(ctx context.Context, i *stats.ConnTagInfo) context.Context {
+	h.tagConnCalled.Store(true)
+	return ctx
+}
+
+func (h *noopStatsHandler) HandleConn(ctx context.Context, s stats.ConnStats) {
+	h.handleConnCalled.Store(true)
+}
+
+// testCredsBundle implements `credentials.Bundle` and `bootstrap.extraDialOptions`.
+// It encapsulates an insecure credential and returns dial options with a mock dialer and a mock
+// stats handler.
+type testCredsBundle struct {
+	credentials.Bundle
+	dialerCalled      *atomic.Bool
```

Comment:
We switched from using `chan struct{}` (in `xds_client_custom_dialer_test.go`) to the atomic package.

Channel are used for communication between goroutines (but it can also be used as mutex/atomic synchronization). In this case, I think we should be OK as we are using it for "one-time" signaling / synchronization between goroutines. Looking at https://go.dev/ref/mem, I don't see any issues.

@easwars what would be the preferred approach here? Would there be any issues switching from channel to atomic? From Go tips/best practices, it is recommended to use channel for one-time signaling.

Reply:
AFAICT channel seems to be the obvious choice if you need to block on the event you expect. But here we can ensure expected events (both the dialer and stats handler) should have happened before we validate as we call `cc.Close()` first (commented). Plus for stats handler atomic simply works better as we expect multiple events but all we want to check is the presence of any.